src/reflect/scala/reflect/internal/util/FreshNameCreator.scala

```diff
@@ -20,8 +20,13 @@ class FreshNameCreator(creatorPrefix: String = "") {
    */
   def newName(prefix: String): String = {
     val safePrefix = NameTransformer.encode(prefix)
-    counters.putIfAbsent(safePrefix, new AtomicLong(0))
-    val idx = counters.get(safePrefix).incrementAndGet()
+    var counter = counters.get(safePrefix)
+    if (counter eq null) {
+      val fresh = new AtomicLong(0)
+      val raced = counters.putIfAbsent(safePrefix, fresh)
+      counter = if (raced eq null) fresh else raced
+    }
+    val idx = counter.incrementAndGet()
```

Comment:
(The lambda won't be allocated each time as it doesn't capture and `LambdaMetaFactory` will cache it.)